source/CCustomOpcodeSystem.cpp

```diff
@@ -399,16 +399,100 @@ namespace CLEO
 	}
 
-	// read string parameter according to convention on strings. Always null terminated
-	char* ReadStringParam(CRunningScript *thread, char* buf, DWORD bufSize)
+	const char* ReadStringParam(CRunningScript *thread, char* buff, DWORD buffSize)
 	{
-		static char internal_buf[MAX_STR_LEN];
-		if (!buf) { buf = internal_buf; bufSize = MAX_STR_LEN; }
+		static char internal_buf[MAX_STR_LEN + 1]; // and terminator
+
+		if (buff == nullptr || buffSize <= 1)
+		{
+			CLEO_SkipOpcodeParams(thread, 1);
+			LOG_WARNING(thread, "Invalid ReadStringParam argument!");
+			return nullptr; // invalid argument
+		}
+
+		// if available obtain pointer to the source, null terminated, string data
+		char* source = nullptr;
+		auto paramType = thread->PeekDataType();
+		auto arrayType = IsArray(paramType) ? thread->PeekArrayDataType() : eArrayDataType::ADT_NONE;
+		auto isVariableInt = IsVariable(paramType) && (arrayType == eArrayDataType::ADT_NONE || arrayType == eArrayDataType::ADT_INT);
+
+		if (IsImmInteger(paramType) || isVariableInt)
+		{
+			source = (char*)CLEO_PeekIntOpcodeParam(thread); // integer address to text buffer
+
+			if ((size_t)source <= CCustomOpcodeSystem::MinValidAddress)
+			{
+				CLEO_SkipOpcodeParams(thread, 1);
+				LOG_WARNING(thread, "Invalid '0x%X' pointer of input string argument #%d in script %s", opcodeParams[0].dwParam, CLEO_GetParamsHandledCount(), ScriptInfoStr(thread).c_str());
+				return nullptr; // error
+			}
+		}
+		else if (IsImmString(paramType))
+		{
+			switch (paramType)
+			{
+				// always terminated in-code strings
+			case DT_TEXTLABEL:
+			case DT_STRING:
+				source = (char*)(thread->GetBytePointer() + sizeof(eDataType)); // skip type info, go to data
+				break;
+
+			case DT_VARLEN_STRING: // never terminated in-code string
+			default:
+				break;
+			}
+		}
+		else if (IsVarString(paramType))
+		{
+			auto str = (char*)CLEO_PeekPointerToScriptVariable(thread);
+			if (str != nullptr)
+			{
+				switch (paramType)
+				{
+					// short string variable
+				case DT_VAR_TEXTLABEL:
+				case DT_LVAR_TEXTLABEL:
+				case DT_VAR_TEXTLABEL_ARRAY:
+				case DT_LVAR_TEXTLABEL_ARRAY:
+					if (strnlen_s(str, 8) < 8) // if all space used by text then there is no terminator
+						source = str;
+					break;
+
+					// long string variable
+				case DT_VAR_STRING:
+				case DT_LVAR_STRING:
+				case DT_VAR_STRING_ARRAY:
+				case DT_LVAR_STRING_ARRAY:
+					if (strnlen_s(str, 16) < 16) // if all space used by text then there is no terminator
+						source = str;
+					break;
+
+				default:
+					break;
+				}
+			}
+		}
+
+		if (source != nullptr && buff == nullptr)
+		{
+			// no need to copy data, as user can not access the internal buffer anyway
+			CLEO_SkipOpcodeParams(thread, 1);
+			return source;
+		}
+
+		if (buff == nullptr)
+		{
+			buff = internal_buf;
+
+			if (buffSize == 0) buffSize = sizeof(internal_buf);
+			else buffSize = min(buffSize, sizeof(internal_buf)); // user requested size limit
+		}

```

Comment:
you can delete both IFs as buff can't be nullptr here

Reply:
Updated, it can.

Reply:
Look, this method is used for all CLEO opcodes. Nullptr is only passed from 0AB1 opcode. Why don't we optimize the logic by requiring the buffer? It will remove quite a few branches from the code, optimizing this code for 99% of calls?

Reply:
There is also 
```
scmFunc->stringParams.emplace_back(ReadStringParam(thread));
```

Reply:
I know, it's the same opcode.

Reply:
Updated.

Reply:
Thanks, now you can delete conditions that check if buff == nullptr